### PR TITLE
EZP-23949: Implement a simple installer for Platform

### PR DESCRIPTION
**Pull request by bdunogier**
> [EZP-23949](http://jira.ez.no/browse/EZP-23949)
> Status: work in progress, open for comments

## Scope
Starting from an eZ Platform with dependencies installed, get it running using a simple installation script.

## Usage example
```
$ php ezpublish/console ezplatform:install
Executing 659 queries from vendor/ezsystems/ezpublish-kernel/data/mysql/schema.sql on database ezplatform
Executing 2054 queries from vendor/ezsystems/ezpublish-kernel/data/cleandata.sql on database ezplatform
Copied /home/bertrand/dev-platform-installer/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/PlatformInstallerBundle/src/Command/../Resources/config_templates/ezpublish.yml.clean to ezpublish/config/ezpublish.yml
Copied /home/bertrand/dev-platform-installer/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/PlatformInstallerBundle/src/Command/../Resources/config_templates/ezpublish_dev.yml to ezpublish/config/ezpublish_dev.yml
Copied /home/bertrand/dev-platform-installer/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/PlatformInstallerBundle/src/Command/../Resources/config_templates/ezpublish_prod.yml to ezpublish/config/ezpublish_prod.yml
```

## Implementation of installation itself
Very crude. Uses simple SQL scripts to get data in, and copies configuration files without changing anything to them.
But it is sufficient for now.

## PlatformInstallerBundle
This new bundle, located in `eZ/Bundle/PlatformInstallerBundle` (namespace `EzSystems\PlatformInstallerBundle`) contains the installer itself: the ezplatform:install command, the `EzSystems\PlatformInstallerBundle\Installer\Installer` interface, and the CleanInstaller in the same namespace.

Installers are registered via the `ezplatform.installer` service tag, with a `type` attribute (demo, demo-clean...).

The `clean` installer is shipped by default, but not supported (it's a leftover from the implementation). The `demo` and `demo-clean` installers are in the DemoBundle itself ()

## TODO
- [x] Merge ezsystems/LegacyBridge#7 and ezsystems/ezpublish-kernel#1180 before this one
- [x] Move removed listener to legacy-bridge, and rebase
- [x] Merge #1180, and rebase this one
- [x] Delete or update the DumpSchema Command
- [x] Rebase and cleanup
- [x] Fix CS
- [x] Figure out the line endings errors...
- [x] Move demo data (at least binaries to another repository), and rebase the binaries out of this one
- [ ] Update clean data to have relevant text for the alpha (can be done as separate PR after)
- [ ] Rename PR title

### CODE REVIEW

**Comment by crevillo** [comment]
First suggestion: replace all that message by a progress bar :)

File: eZ/Bundle/PlatformInstallerBundle/src/Command/DumpSchemaCommand.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public function execute(InputInterface $input, OutputInterface $output)
+    {
+        $connection = $this->getContainer()->get('database_connection');
+        $sql = $connection
+            ->getSchemaManager()
+            ->createSchema()
+            ->toSql($connection->getDatabasePlatform());
+        print_r($sql);
```
Comment (crevillo) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/DumpSchemaCommand.php]: i guess we'll dete this print_r :)

Comment (bdunogier) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/DumpSchemaCommand.php]: Yep. I'll probably delete or rewrite this Command altogether anyway :-)

File: eZ/Bundle/EzPublishCoreBundle/EventListener/RequestEventListener.php [hunk 2]
```diff
@@ -58,7 +58,6 @@ public static function getSubscribedEvents()
     {
         return array(
             KernelEvents::REQUEST => array(
-                array( 'onKernelRequestSetup', 190 ),
```
Comment (lolautruche) [inline eZ/Bundle/EzPublishCoreBundle/EventListener/RequestEventListener.php]: Mmm why removing it? Did you move this to legacy bridge?

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            exit(self::EXIT_PARAMETERS_NOT_FOUND);
+        }
+    }
+
+    private function importSchema()
+    {
+        // @todo Should come from a different place
+        // @todo Should be part of the LSE
+        $this->runQueriesFromFile(
+            'vendor/ezsystems/ezpublish-kernel/data/mysql/schema.sql'
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Should rely at least on container parameter.

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        );
+    }
+
+    private function importCleanData()
+    {
+        // @todo Should come from a different place
+        // @todo Should be part of the LSE
+        $this->runQueriesFromFile(
+            'vendor/ezsystems/ezpublish-kernel/data/cleandata.sql'
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Same as above

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    const EXIT_DATABASE_NOT_FOUND_ERROR = 3;
+    const EXIT_GENERAL_DATABASE_ERROR = 4;
+    const EXIT_PARAMETERS_NOT_FOUND = 5;
+    const EXIT_UNKNOWN_INSTALL_TYPE = 6;
+    const EXIT_MISSING_PERMISSIONS = 7;
+
+    protected function configure()
+    {
+        $this->setName( 'ezplatform:install' );
+        $this->addArgument( 'type', InputArgument::REQUIRED, "The type of install, one of: clean, demo-clean" );
```
Comment (andrerom) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: missing demo

**Comment by andrerom** [comment]
Nice to see it progressing :) However as said, not a big fan of putting the binaries/assets in our repos, they are 2-5mb or more, same as discussion on Demo Bundle, if they are committed once (even if just in this PR and removed in later commit), then the repo gets that extra size forever.

**Comment by andrerom** [comment]
Seems like the sql dump for demo-clean and demo is not right, much smaller than cleandata.

**Comment by bdunogier** [comment]
> Seems like the sql dump for demo-clean and demo is not right, much smaller than cleandata.

The demo one should be fixed now. Got confused with my installations, databases, dumps... :-)

**Comment by bdunogier** [comment]
> However as said, not a big fan of putting the binaries/assets in our repos, they are 2-5mb or more. Same as discussion on Demo Bundle, if they are committed once (even if just in this PR and removed in later commit), then the repo gets that extra size forever.

Don't worry, the commit says "remove me", and it will be rebased out, not just removed. I'm leaning towards a separate repo+dependency we would add to the demobundle itself for now. That way, we are free to ditch it at a later point.

**Comment by bdunogier** [comment]
*warning, the branch has been rebased*

**Comment by bdunogier** [comment]
I can't see the line endings errors that are reported on my side... if somebody figures those out, I wouldn't mind a commit :p

**Comment by bdunogier** [comment]
Demo install images moved to http://github.com/bdunogier/demobundle-data. Temporary name of course.

**Comment by andrerom** [comment]
To fix the `\r` issue you can use regex find/replace in your editor, or try to remove the new lines and make new once.

File: eZ/Publish/API/Repository/Values/User/Policy.php [hunk 6]
```diff
@@ -18,7 +18,7 @@
  * @property-read mixed $roleId the role id this policy belongs to
  * @property-read string $module Name of module, associated with the Policy
  * @property-read string $function  Name of the module function Or all functions with '*'
- * @property-read array $limitations an array of \eZ\Publish\API\Repository\Values\User\Limitation
+ * @property-read \eZ\Publish\API\Repository\Values\User\Limitation[] $limitations an array of \eZ\Publish\API\Repository\Values\User\Limitation
```
Comment (andrerom) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: Should probably not be part of this PR?

File: eZ/Publish/API/Repository/Values/User/Role.php [hunk 7]
```diff
@@ -16,7 +16,7 @@
  *
  * @property-read mixed $id the internal id of the role
  * @property-read string $identifier the identifier of the role
- * @property-read array $policies an array of the policies {@link \eZ\Publish\API\Repository\Values\User\Policy} of the role.
+ * @property-read \eZ\Publish\API\Repository\Values\User\Policy[] $policies an array of the policies of the role.
```
Comment (andrerom) [inline eZ/Publish/API/Repository/Values/User/Role.php]: same here?

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    const EXIT_GENERAL_DATABASE_ERROR = 4;
+    const EXIT_PARAMETERS_NOT_FOUND = 5;
+    const EXIT_UNKNOWN_INSTALL_TYPE = 6;
+    const EXIT_MISSING_PERMISSIONS = 7;
+
+    protected function configure()
+    {
+        $this->setName( 'ezplatform:install' );
+
+        $this->addArgument( 'type', InputArgument::REQUIRED, "The type of install" );
```
Comment (andrerom) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: There is now no help text describing which options exists, or?
Besides we should default to clean or?

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    private function configuredDatabaseExists()
+    {
+        $this->db = $this->getContainer()->get( 'database_connection' );
+        try
+        {
+            $this->db->connect();
+        }
+        catch ( ConnectionException $e )
+        {
+            // @todo 1049 is MySQL's code for "database doesn't exist", refactor
```
Comment (pspanja) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: `SchemaManager::listDatabases()` might be of use.

File: eZ/Bundle/PlatformInstallerBundle/src/DependencyInjection/EzSystemsPlatformInstallerExtension.php [hunk 10]
```diff
@@ -0,0 +1,19 @@
+<?php
+namespace EzSystems\PlatformInstallerBundle\DependencyInjection;
```
Comment (pspanja) [inline eZ/Bundle/PlatformInstallerBundle/src/DependencyInjection/EzSystemsPlatformInstallerExtension.php]: Missing licence.

File: eZ/Bundle/PlatformInstallerBundle/src/EzSystemsPlatformInstallerBundle.php [hunk 11]
```diff
@@ -0,0 +1,15 @@
+<?php
+namespace EzSystems\PlatformInstallerBundle;
```
Comment (pspanja) [inline eZ/Bundle/PlatformInstallerBundle/src/EzSystemsPlatformInstallerBundle.php]: Missing licence.

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php (hunk 8 above)
Comment (bdunogier) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: > There is now no help text describing which options exists, or?
Hah, funny that you mention this one. Thing is, they were hardcoded before, but now they depend on the service tag and container. But the container is not available in `configure()`. The easiest would be an option (--list-types) that lists the available installation types.

And `clean` is in fact the only option that is always available (without the DemoBundle that is).

File: eZ/Publish/API/Repository/Values/User/Policy.php (hunk 6 above)
Comment (bdunogier) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: Well, it's only phpdoc, and the commit message isn't linked to this PR. Figured it wouldn't do much harm. But it can be extracted safely of course.

**Comment by bdunogier** [comment]
Thank you @pspanja.

File: eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php [hunk 12]
```diff
@@ -68,4 +68,43 @@ public static function dumpAssetsHelpText( CommandEvent $event )
...
+    ___/\\\/////\\\_______/\\\/_________________\/\\\/////////_______\/\\\____\////////\\\___\////\\\////__\////\\\//______/\\\///\\\__\/\\\/////\\\__/\\\///\\\\\///\\\_
+     __/\\\\\\\\\\\______/\\\/___________________\/\\\________________\/\\\______/\\\\\\\\\\_____\/\\\_________\/\\\_______/\\\__\//\\\_\/\\\___\///__\/\\\_\//\\\__\/\\\_
+      _\//\\///////_____/\\\/_____________________\/\\\________________\/\\\_____/\\\/////\\\_____\/\\\_/\\_____\/\\\______\//\\\__/\\\__\/\\\_________\/\\\__\/\\\__\/\\\_
+       __\//\\\\\\\\\\__/\\\\\\\\\\\\\\\___________\/\\\______________/\\\\\\\\\_\//\\\\\\\\/\\____\//\\\\\______\/\\\_______\///\\\\\/___\/\\\_________\/\\\__\/\\\__\/\\\_
+        ___\//////////__\///////////////____________\///______________\/////////___\////////\//______\/////_______\///__________\/////_____\///__________\///___\///___\///__
+EOT
+        );
+        $event->getIO()->write( '' );
+        $event->getIO()->write( '' );
+        $event->getIO()->write( '<fg=cyan>Welcome to eZ Platform alpha!</fg=cyan>' );
```
Comment (lolautruche) [inline eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php]: Let's not hardcode `alpha` :wink:

File: eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php [hunk 13]
```diff
@@ -68,4 +68,43 @@ public static function dumpAssetsHelpText( CommandEvent $event )
...
+        $event->getIO()->write( 'A. Installing eZ Platform purly using command:' );
+        $event->getIO()->write( '<comment>    $ php ezpublish/console ezplatform:install clean</comment>' );
+        $event->getIO()->write( '' );
+        $event->getIO()->write( 'B. Installing legacy bridge to use eZ Publish is done in following way:' );
+        $event->getIO()->write( '<comment>    $ php composer.phar require ezsystems/legacy-bridge:~2015.01</comment>' );
+        $event->getIO()->write( '' );
+        $event->getIO()->write( '<options=underscore>After executing one, or both of those. You can launch your browser* and get started on your publishing journey!</options=underscore>' );
+        $event->getIO()->write( '' );
+        $event->getIO()->write( '' );
+        $event->getIO()->write( '* Assuming you have configured your web server (Apache/Nginx) correctly' );
```
Comment (lolautruche) [inline eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php]: Why not getting everything in one single heredoc block ?

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php (hunk 8 above)
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Then you should do a *command as a service*

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        $this->checkPermissions();
+        $this->checkParameters();
+        $this->checkDatabase();
+
+        $type = $input->getArgument( 'type' );
+        $installer = $this->getInstaller( $type );
+        if ( $installer === false )
+        {
+                $output->writeln( "Unknown install type '$type'" );
+                exit( self::EXIT_UNKNOWN_INSTALL_TYPE );
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Too many spaces I guess :)

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php (hunk 8 above)
Comment (bdunogier) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Added `ezplatform:list-install-types` command for that. I'd rather not go for a default now, as we'd have to hardcode it to `clean`, and we don't really "support" clean (no default bundle, for instance).

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+            exit( self::EXIT_PARAMETERS_NOT_FOUND );
+        }
+    }
+
+    /**
+     * @throws \Exception if an unexpected database error occurs
+     */
+    private function configuredDatabaseExists()
+    {
+        $this->db = $this->getContainer()->get( 'database_connection' );
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Note that this will only retrieve the default one.

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+     */
+    private function getInstaller( $type )
+    {
+        $installers = $this->getContainer()->getParameter( 'ezplatform.installers' );
+        if ( !isset( $installers[$type] ) )
+        {
+            return false;
+        }
+
+        return $this->getContainer()->get( $installers[$type] );
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Considering you modify this command to be a service, you could add installers with a setter (e.g. `addInstaller()`) in your compiler pass.
You may keep `ezplatform.installers` if needed.

File: eZ/Bundle/PlatformInstallerBundle/src/Command/ListInstallTypesCommand.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+use Doctrine\DBAL\Exception\ConnectionException;
+use EzSystems\PlatformInstallerBundle\Installer\Installer;
+use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ListInstallTypesCommand extends ContainerAwareCommand
+{
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/ListInstallTypesCommand.php]: This command would be useless if the other one was a service.

File: eZ/Bundle/PlatformInstallerBundle/src/Installer/DbBasedInstaller.php [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+                    count( $queries ),
+                    $file,
+                    $this->db->getDatabase()
+                )
+            );
+        }
+
+        foreach ( $queries as $query )
+        {
+            $this->db->exec( $query );
```
Comment (lolautruche) [inline eZ/Bundle/PlatformInstallerBundle/src/Installer/DbBasedInstaller.php]: Progress bar here maybe?

File: eZ/Bundle/PlatformInstallerBundle/src/Resources/config_templates/clean/ezpublish.yml [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        list:
+            - default
+        # Siteaccess groups. Use them to group common settings.
+        groups:
+            site_group: [site]
+        default_siteaccess: site
+        match:
+            URIElement: 1
+
+    imagemagick:
```
Comment (crevillo) [inline eZ/Bundle/PlatformInstallerBundle/src/Resources/config_templates/clean/ezpublish.yml]: imagemagick? still needed?

File: eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php (hunk 15 above)
Comment (bdunogier) [inline eZ/Bundle/PlatformInstallerBundle/src/Command/InstallPlatformCommand.php]: Well, the default configuration uses the default database connection. It's not idea, but it's not supposed to be (yet) :-)

File: eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php (hunk 13 above)
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php]: would the tags work using that?

Comment (bdunogier) [inline eZ/Bundle/EzPublishCoreBundle/Composer/ScriptHandler.php]: No reason why they would not, it's a string, is it not ? Trying.

File: eZ/Publish/API/Repository/Values/User/Policy.php (hunk 6 above)
Comment (andrerom) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: it's a API change mister ;)

**Comment by crevillo** [comment]
Open question @bdunogier and all . Now legacy is being decoupled, how about removing some unneded tables from the install process? Maybe is a bit too early to do it because customers are still relying on legacy for the admin part?

File: eZ/Publish/API/Repository/Values/User/Policy.php (hunk 6 above)
Comment (bdunogier) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: I'll revert them.

**Comment by andrerom** [comment]
@crevillo That would be a later followup, think we first will try to come up with a data format instead of using raw sql, and a more transportable schema format for use with Doctrine DBAL.

**Comment by andrerom** [comment]
+1 if travis agrees :)

**Comment by bdunogier** [comment]
> if travis agrees :) 

I'd be surprised if it did...

**Comment by andrerom** [comment]
seems he did, guess you where thinking about bdd tests.

**Comment by bdunogier** [comment]
> seems he did, guess you where thinking about bdd tests.

I was. Too many failures lately :-)
